File: Problems/Algorithms/703. Kth Largest Element in a Stream/KthLargest.kt
class KthLargest(k: Int, nums: IntArray) {
    
    private val k: Int
    private val heap: PriorityQueue<Int>
    init {
        this.k = k
        heap = PriorityQueue<Int>()
        
        for (num in nums) {
            heap.offer(num)
        }
        
        while (heap.size > k) {
            heap.poll()
        }
    }

    fun add(`val`: Int): Int {
        heap.offer(`val`)
        
        if (heap.size > k) {
            heap.poll()
        }
        
        return heap.peek()
    }

}

/**
 * Your KthLargest object will be instantiated and called as such:
 * var obj = KthLargest(k, nums)
 * var param_1 = obj.add(`val`)
 */

File: Problems/Algorithms/703. Kth Largest Element in a Stream/kth-largest.cpp
#include <vector>
#include <queue>

using namespace std;

class KthLargest {
private:
    priority_queue<int, vector<int>, greater<int>> minHeap;
    int k;

public:
    KthLargest(int k, vector<int>& nums) {
        this->k = k;

        for (int num : nums) {
            add(num);
        }
    }
    
    int add(int val) {
        if (minHeap.size() < k || minHeap.top() < val) {
            minHeap.push(val);
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }
        
        return minHeap.top();
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */